crates/adapters/src/format/avro/schema.rs

```diff
@@ -0,0 +1,224 @@
+        .map_err(|e| format!("error validating map value schema: {e}"))?;
+
+    Ok(())
+}
+
+/// Check that Avro schema can be deserialized as SQL `TIMESTAMP` type.
+fn validate_timestamp_schema(avro_schema: &AvroSchema) -> Result<(), String> {
+    // TODO: we can support TimestampMillis by transforming them to micros on the fly.
+    if avro_schema == &AvroSchema::TimestampMillis {
+        return Err("Avro timestamp encoding using 'timestamp-millis' type is currently not supported; use 'timestamp-micros instead'".to_string());
+    }
+
+    if avro_schema != &AvroSchema::TimestampMicros && avro_schema != &AvroSchema::Long {
+        return Err(format!(
+            "invalid Avro schema for a column of type 'TIMESTAMP': expected 'timestamp-micros' or 'long', but found {}",
+            schema_json(avro_schema)
+        ));
+    }
+
+    Ok(())
+}
+
+/// Check that Avro schema can be deserialized as SQL `TIME` type.
+fn validate_time_schema(avro_schema: &AvroSchema) -> Result<(), String> {
+    // TODO: we can support TimeMillis by transforming them to micros on the fly.
+    if avro_schema == &AvroSchema::TimeMillis {
+        return Err("Avro time encoding using 'time-millis' type is currently not supported; use 'time-micros instead'".to_string());
+    }
+
+    if avro_schema != &AvroSchema::TimeMicros && avro_schema != &AvroSchema::Long {
+        return Err(format!(
+            "invalid Avro schema for a column of type 'TIME': expected 'time-micros' ot 'long', but found {}",
+            schema_json(avro_schema)
+        ));
+    }
+
+    Ok(())
+}
+
+/// Check that Avro schema can be deserialized a SQL column with the given
+/// column type.
+pub fn validate_field_schema(
+    avro_schema: &AvroSchema,
+    field_schema: &ColumnType,
+) -> Result<(), String> {
+    if field_schema.nullable {
+        let avro_inner = nullable_schema_value_schema(avro_schema).ok_or_else(|| {
+            format!(
+                "nullable column with non-nullable Avro schema {}",
+                schema_json(avro_schema)
+            )
+        })?;
+        let mut field_schema = field_schema.clone();
+        field_schema.nullable = false;
+        return validate_field_schema(avro_inner, &field_schema);
+    };
+
+    let expected = match field_schema.typ {
+        SqlType::Boolean => AvroSchema::Boolean,
+        SqlType::TinyInt | SqlType::SmallInt | SqlType::Int => AvroSchema::Int,
+        SqlType::BigInt => AvroSchema::Long,
+        SqlType::Real => AvroSchema::Float,
+        SqlType::Double => AvroSchema::Double,
+        SqlType::Decimal => {
+            return Err("not implemented: Avro deserialization for the 'DECIMAL' type".to_string());
+        }
+        SqlType::Char | SqlType::Varchar => AvroSchema::String,
+        SqlType::Binary | SqlType::Varbinary => AvroSchema::Bytes,
+        SqlType::Time => {
+            return validate_time_schema(avro_schema);
+        }
+        SqlType::Date => AvroSchema::Int,
```

Comment:
no date in Avro?
weird

Reply:
Actually there is, but it's not always used apparently. Good catch, I added that.